File: dcp-solutions/2024-10-20/solution.rb
```ruby
def findItinerary(flights, start)
  graph = flights.group_by { |s, d| s }.map { |_, destinations| [destinations.first, destinations.map(&:last)] }
  
  def dfs(cur, path)
    return [] if cur == start
    path << cur
    graph[cur].sort_by! { |d| -1 } # sort destinations in reverse lexicographical order
    
    while not graph[cur].empty? and graph[cur][0][0] != start
      result = dfs(graph[cur][0][0], path)
      return result if result
      graph[cur].shift
    end
    
    if graph[cur].empty?
      path.pop # backtracking
      []
    else
      d1, d2 = graph[cur].first
      result = [d1] + dfs(d1, path) + [d2]
      return result
    end
  end
  
  if graph[start].empty? and start == flights.map { |s, _| s }.uniq[0]
    return nil
  end
  
  result = dfs(start, [])
  
  unless result
    # no valid itinerary exists
    return []
  end
  
  # lexicographically smallest itinerary in reverse order
  result.reverse!
end
```

File: dcp-solutions/2024-10-20/solution.js
```javascript
function findItinerary(flights) {
    flights.sort((a, b) => a[0].localeCompare(b[0]));
    const graph = {};
    for (const [from, to] of flights) {
        if (!graph[from]) graph[from] = [];
        graph[from].push(to);
    }
    
    function dfs(currentCity, path) {
        while (graph[currentCity] && graph[currentCity][0] < path[path.length - 1]) {
            dfs(graph[currentCity][0], [...path, graph[currentCity][0]]);
        }
        
        if (graph[currentCity] && graph[currentCity].length === 0) return [currentCity, ...path];
        if (!graph[currentCity]) return null;
        
        for (const nextCity of graph[currentCity]) {
            const result = dfs(nextCity, [...path, nextCity]);
            if (result !== null) return result;
        }
    }
    
    return dfs(flights[0][0], [flights[0][0]]) || null;
}
```